node/src/main/scala/coop/rchain/node/web/Transaction.scala

```diff
@@ -0,0 +1,257 @@
+
+final case class TransactionInfo(transaction: Transaction, transactionType: TransactionType)
+final case class TransactionResponse(data: List[TransactionInfo])
+
+trait TransactionAPI[F[_]] {
+  def getTransaction(blockHash: Blake2b256Hash): F[List[TransactionInfo]]
+}
+
+/**
+  * This API is totally based on how RevVault.rho is written. If the `RevVault.rho` is re-written or changed,
+  * this API might end up with useless.
+  */
+final case class TransactionAPIImpl[F[_]: Concurrent](
+    blockReportAPI: BlockReportAPI[F],
+    // The transferUnforgeable can be retrieved based on the deployer and the timestamp of RevVault.rho
+    // in the genesis ceremony.
+    transferUnforgeable: Par
+) extends TransactionAPI[F] {
+  // Helper which accepts TransactionType constructor
+  private def makeDeployTransaction(
+      d: DeployInfoWithEventData,
+      report: SingleReport,
+      txCtor: String => TransactionType
+  ) = findTransactions(report).map(t => TransactionInfo(t, txCtor(d.deployInfo.get.sig)))
+
+  def getTransaction(blockHash: Blake2b256Hash): F[List[TransactionInfo]] =
+    blockReportAPI.blockReport(blockHash.toByteString, forceReplay = false).map {
+      case Right(b) =>
+        val userTransactions = b.deploys.flatMap(d => {
+          d.report
+            .zip(Seq(PreCharge, UserDeploy, Refund))
+            .flatMap {
+              case (report, txCtor) =>
+                makeDeployTransaction(d, report, txCtor)
+            }
+        })
+        val systemDeployTransactions = b.systemDeploys.flatMap { s =>
+          // system doesn't get precharge and refund , so it would always get one
+          val transactions = findTransactions(s.report.head)
+          val txCtor = s.systemDeploy.get.systemDeploy.value match {
+            case SlashSystemDeployDataProto(_, _)  => SlashingDeploy
+            case CloseBlockSystemDeployDataProto() => CloseBlock
+          }
+          transactions.map(t => TransactionInfo(t, txCtor(Base16.encode(blockHash.bytes.toArray))))
+        }
+        (userTransactions ++ systemDeployTransactions).toList
+
+      case Left(_) => List.empty
+    }
+
+  private def findTransactions(report: SingleReport): Seq[Transaction] = {
+    val transactions = report.events
+      .map(_.report.value)
+      .collect {
+        case ReportCommProto(Some(consume), produces) =>
+          consume.channels.headOption.map((_, produces))
+      }
+      .flatten
+      .collect { case (channel, produces) if channel == transferUnforgeable => produces.head }
+      .map { produce =>
+        val fromAddr       = produce.data.get.pars.head.exprs.head.getGString
+        val toAddr         = produce.data.get.pars(2).exprs.head.getGString
+        val amount         = produce.data.get.pars(3).exprs.head.getGInt
+        val retUnforgeable = produce.data.get.pars(5)
+        Transaction(fromAddr, toAddr, amount, retUnforgeable, None)
+      }
```

Comment:
Without converting to concrete structure like `toList` here, it will execute the whole composition from where the iterator is created.

First iteration is here. To existing iterator `transactions`, another map will be added to create a Set.
```scala
transactions.map(_.retUnforgeable).toSet
```

Second iteration is here. To existing iterator `transactions` another map will be added to produce a List.
```scala
transactions.map { t =>
  val failReason = failedMap.getOrElse(t.retUnforgeable, None)
  t.copy(failReason = failReason)
}.toList
```

Because iterators are lazy, they will run only when converted to concrete collection type like `toList` (Scala 3 `to(List)`). This is the same concept used in fs2.Stream which is run with `to(List)` or `drain`.